File: 684.redundant-connection.go
func findRedundantConnection(edges [][]int) []int {
	N := len(edges)

	var parent = make([]int, N+1)
	var size = make([]int, N+1)

	uf_init(parent, size)

	var res = make([]int, 2)

	for i := 0; i < N; i++ {
		x := edges[i][0]
		y := edges[i][1]

		if uf_connected(parent, x, y) {
			res[0] = x
			res[1] = y
		} else {
			uf_add(parent, size, x, y)
		}
	}

	return res
}


func uf_init(parent []int, size []int) {
	for i := 0; i < len(parent); i++ {
		parent[i] = i
	}

	for i := 0; i < len(size); i++ {
		size[i] = 1
	}
}

func uf_add(parent []int, size []int, x int, y int) {
	i := uf_find(parent, x)
	j := uf_find(parent, y)

	if i == j {
		return
	}

	if size[i] > size[j] {
		parent[j] = i
		size[i] += size[j]
	} else {
		parent[i] = j
		size[j] += size[i]
	}
}

func uf_find(parent []int, x int) int {
	for x != parent[x] {
		x = parent[x]
	}

	return x
}

func uf_connected(parent []int, x int, y int) bool {
	return uf_find(parent, x) == uf_find(parent, y)
}

File: 684.redundant-connection.php
class Solution {

    private $parent;

    function find($x)
    {
        if (!isset($this->parent[$x])) {
            $this->parent[$x] = $x;
        }

        if ($this->parent[$x] != $x) {
            return $this->find($this->parent[$x]);
        }
        return $this->parent[$x];
    }

    /**
     * @param Integer[][] $edges
     * @return Integer[]
     */
    function findRedundantConnection($edges) {
        foreach ($edges as $edge) {
            list($x, $y) = $edge;
            
            $parentX = $this->find($x);
            $parentY = $this->find($y);
            
            if ($parentY != $parentX) {
                $this->parent[$parentY] = $parentX;
            } else {
                return $edge;
            }
        }
    }
}
